Enable S3 backup

JacieChao opened the pull request:
To support S3 backup for longhorn-manager, some components need aws credential.
- User need to create credential secret manually, and set secret name to `Setting.secretConfig`
- If backupTarget for s3, longhorn-manager need to get credential and passing them as an environment variable to engine.

JacieChao commented on types/types.go:
```diff
@@ -52,7 +55,7 @@ const (
 	// 4. cronjob pod suffix is 11
 	// 5. Dash and buffer for 2
-	MaximumJobNameSize = 8
+	MaximumJobNameSize = 30
```
I'm not pretty sure how to calculate this const variable. When I'm testing `Recurring backup and snapshot` from UI, it returns an error because this value is too short for a CronJob name. So I change it to a random value this time. 
I would be appreciate for your help.

yasker replied:
Use new version of UI: rancher/longhorn-ui:6d74dfc , see https://github.com/rancher/longhorn-manager/commit/1e5f85cfe8cc02112a690510b51564697fd18f14

yasker replied:
And separate this to another commit next time.

yasker commented on api/setting.go:
```diff
@@ -37,6 +37,8 @@ func (s *Server) SettingsGet(w http.ResponseWriter, req *http.Request) error {
 		value = si.DefaultEngineImage
 	case types.SettingEngineUpgradeImage:
 		value = si.EngineUpgradeImage
+	case types.SettingSecretConfig:
```
The name `SecretConfig` is too ambiguous. Use `BackupTargetCredentialSecret` instead.

yasker commented on app/volume.go:
```diff
@@ -238,7 +256,12 @@ func (job *Job) backupAndCleanup() error {
 	}
-	if err := job.engine.SnapshotBackup(job.snapshotName, job.backupTarget, job.labels); err != nil {
+	secret := make(map[string]string)
+	if strings.Contains(job.backupTarget, "s3") {
```
Same mistake as the engine one here.

yasker commented on controller/replica_controller.go:
```diff
@@ -389,6 +390,36 @@ func (rc *ReplicaController) CreatePodSpec(obj interface{}) (*v1.Pod, error) {
 			},
 		}
 	}
+	if r.Spec.RestoreName != "" && r.Spec.RestoreFrom != "" && strings.Contains(r.Spec.RestoreFrom, "s3") {
```
Same mistake using `strings.Contains`

yasker commented on controller/volume_controller.go:
```diff
@@ -880,6 +881,9 @@ func (vc *VolumeController) createCronJob(v *longhorn.Volume, job *types.Recurri
 	}
 	if job.Type == types.RecurringJobTypeBackup {
 		cmd = append(cmd, "--backuptarget", backupTarget)
+		if backupTarget != "" && strings.Contains(backupTarget, "s3") {
```
Don't show keys here in the cmdline. It will be shown in the CronJob spec. Use secret and mount the secret instead.

yasker commented on engineapi/engine.go:
```diff
@@ -159,3 +159,16 @@ func (e *Engine) Upgrade(binary string, replicaURLs []string) error {
+			return errors.Errorf("invalid credential for s3 backup")
+		}
+		// test credential connect
+		if _, err := util.Execute(LonghornEngineBinary, "validate", "s3", "--dest", backupTarget, "--accessKey", secret[types.SecretAccessKey], "--secretKey", secret[types.SecretKey]); err != nil {
```
Just call list backup for that target. Engine should return empty list.

yasker commented on engineapi/snapshot.go:
```diff
@@ -78,15 +79,18 @@ func (e *Engine) SnapshotPurge() error {
 }
 
-func (e *Engine) SnapshotBackup(snapName, backupTarget string, labels map[string]string) error {
+func (e *Engine) SnapshotBackup(snapName, backupTarget string, labels map[string]string, secret map[string]string) error {
```
`credential`, not `secret`.

`Secret` is where Kubernetes keeps encrypted data.

`Credential` is the thing we need to access something, S3 in this case.

yasker commented on engineapi/snapshot.go:
```diff
@@ -78,15 +79,18 @@ func (e *Engine) SnapshotPurge() error {
+	if strings.Contains(backupTarget, "s3") && (secret == nil || secret[types.SecretAccessKey] == "" || secret[types.SecretKey] == "") {
+		return errors.Errorf("could not backup snapshot '%s' without aws credential", snapName)
+	}
+	args := []string{"--url", e.cURL, "backup", "--accessKey", secret[types.SecretAccessKey], "--secretKey", secret[types.SecretKey], "create", "--dest", backupTarget}
```
I don't think we need new options `--accessKey` and `--secretKey` for the engine. We can set them in the environment variable before executing the command, and Longhorn Engine can pick them up explicitly.

The problems with setting them in the cmdline are:
1. if the command failed, the log will spit out the credential in plaintext. Though it's not a deal-breaker.
2. Every time we are adding a new type of destination, we will add some more parameters to the engine in order to support it. I definitely prefer the credential to be passed implicitly instead.

yasker commented on types/types.go:
```diff
@@ -28,6 +28,9 @@ const (
 	EnvPodNamespace = "POD_NAMESPACE"
 	EnvPodIP        = "POD_IP"
 
+	SecretAccessKey = "access-key"
```
`AWS_SECRET_ACCESS_KEY` and `AWS_ACCESS_KEY_ID`

yasker commented on types/types.go:
```diff
@@ -28,6 +28,9 @@ const (
 	EnvPodNamespace = "POD_NAMESPACE"
 	EnvPodIP        = "POD_IP"
 
+	SecretAccessKey = "AWS_ACCESS_KEY_ID"
```
`AWSxxxx = xxxx`.

Btw, the naming of access key is inconsistent with the one for the engine code. And the `SecretAccessKey` is not the access key, it's the secret key.

It's easier if we just stylize the AWS environment variable name and keep them consistent.

yasker commented on app/volume.go:
```diff
@@ -238,7 +238,8 @@ func (job *Job) backupAndCleanup() error {
 		return err
 	}
-	if err := job.engine.SnapshotBackup(job.snapshotName, job.backupTarget, job.labels); err != nil {
+
```
Better to add a comment here noted that the cronjob template covered the credential already.

yasker commented on controller/volume_controller.go:
```diff
@@ -968,6 +968,35 @@ func (vc *VolumeController) createCronJob(v *longhorn.Volume, job *types.Recurri
+	if job.Type == types.RecurringJobTypeBackup {
+		if credentialSecret != "" {
+			sourceEnv := cronJob.Spec.JobTemplate.Spec.Template.Spec.Containers[0].Env
+			accessKeyEnv := v1.EnvVar{
```
Same code as in the backup restore case. We can create a function for it.

yasker commented on controller/volume_controller.go:
```diff
@@ -995,13 +1024,15 @@ func (vc *VolumeController) updateRecurringJobs(v *longhorn.Volume) (err error)
 		return err
 	}
+
+	settings, err := vc.ds.GetSetting()
```
Err is left unchecked and no checking for setting to be nil.

yasker commented on engineapi/backups.go:
```diff
@@ -86,8 +88,12 @@ func parseOneBackup(output string) (*Backup, error) {
-func (b *BackupTarget) ListVolumes() ([]*BackupVolume, error) {
+func (b *BackupTarget) ListVolumes(credential map[string]string) ([]*BackupVolume, error) {
 	args := []string{"backup", "ls", "--volume-only", b.url}
+	if credential != nil {
```
If it's for s3.

These codes should be made into a function.

yasker commented on engineapi/engine.go:
```diff
@@ -159,3 +160,20 @@ func (e *Engine) Upgrade(binary string, replicaURLs []string) error {
 	return nil
 }
+
+func ValidateCredential(backupTarget string, credential map[string]string) error {
```
ListVolume should already cover this?

yasker commented on manager/engine.go:
```diff
@@ -215,3 +239,15 @@ func (m *VolumeManager) DeleteBackup(backupName, volumeName string) error {
 	return engineapi.DeleteBackup(url)
 }
+
+func (m *VolumeManager) ValidateCredential() error {
```
Where has this function been used?

yasker commented:
LGTM. Will merge this after engine part is merged, since this need to include an updated engine image to enable the feature.

Also please write a bit about how to use this feature in the comment, e.g. create the secret, set the setting etc.

yasker commented:
`rancher/longhorn-engine:64052c3` has been made available for this PR.

JacieChao commented:
We need to do several configurations to enable S3 backup. (https://github.com/rancher/longhorn/issues/52)
- Create a credential secret
For now, we need to create credential secret manually.
```
apiVersion: v1
kind: Secret
metadata:
  name: jacie-s3-secret
type: Opaque
data:
  AWS_ACCESS_KEY_ID: your-access-key-with-base64-encode
  AWS_SECRET_ACCESS_KEY: your-secret-key-with-base64-encode
```
Then create your credential secret.
```
$ kubectl create -f ./secret.yaml
secret "mysecret" created
```
- Set credential secret for longhorn setting.
After setting s3 backupTarget from UI, for now, we need to edit longhorn setting manually and save your credential secret to Setting.BackupTargetCredentialSecret.
```
$ kubectl edit setting longhorn-manager-settings -n longhorn-system
setting.longhorn.rancher.io "longhorn-manager-settings" edited
```
- Try S3 backup from UI.

yasker commented:
@JacieChao In fact, the UI should automatically generate the setting for the secret.

You need to update https://github.com/rancher/longhorn-manager/blob/master/api/model.go#L280

yasker commented:
@JacieChao Can you rebase the code as well?

yasker commented on engineapi/backups.go:
```diff
@@ -86,8 +86,12 @@ func parseOneBackup(output string) (*Backup, error) {
 }
 
-func (b *BackupTarget) ListVolumes() ([]*BackupVolume, error) {
+func (b *BackupTarget) ListVolumes(credential map[string]string) ([]*BackupVolume, error) {
```
I just realize we can put credential inside the BackupTarget. Let's refactor it in after this patch gets merged.

yasker replied:
But why only List and Get get filled in? How about List(), GetBackup() and DeleteBackup()?

JacieChao commented:
@yasker  I have rebased code and fixed problems. But longhorn-engine code need to update as well since we set credential into environment variable in a CronJob, but I still check credential in map which should be nil when backup by CronJob. I will create a new PR to longhorn-engine.

yasker commented:
@JacieChao No problem.

Let's wait a bit for the test to be ready, before merge in all the stuff. The test should catch all those issues.

yasker commented on engineapi/backups.go:
```diff
@@ -15,8 +15,9 @@ import (
-	Image string
+	URL              string
+	Image            string
+	CredentialSecret map[string]string
```
`Credential`, not `CredentialSecret`.

yasker commented on engineapi/backups.go:
```diff
@@ -28,10 +29,11 @@ type backupVolume struct {
 }
 
-func NewBackupTarget(backupTarget, engineImage string) *BackupTarget {
+func NewBackupTarget(backupTarget, engineImage string, credentialSecret map[string]string) *BackupTarget {
```
`credential`.

yasker commented on datastore/longhorn.go:
```diff
@@ -50,6 +50,19 @@ func (s *DataStore) GetSetting() (*longhorn.Setting, error) {
 	return result, err
 }
 
+func (s *DataStore) GetCredentialSecret(secretName string) (map[string]string, error) {
```
It's better to call it `GetCredentialFromSecret`

Secret is a k8s object to store secret things. Credential is the key to some kind of service. CredentialSecret is the k8s secret which stored the credential.

yasker commented on util/util.go:
```diff
@@ -295,3 +299,67 @@ func SplitStringToMap(str, separator string) map[string]struct{} {
+	if err != nil {
+		return err
+	}
+	if backupType == "s3" {
```
Make `s3` a const. 

`types.BackupStoreTypeS3 = "s3"`

yasker commented on util/util.go:
```diff
@@ -295,3 +299,67 @@ func SplitStringToMap(str, separator string) map[string]struct{} {
+	}
+	if backupType == "s3" {
+		// environment variable has been set in cronjob
+		if credentialSecret == nil && (os.Getenv(AWSAccessKey) != "" && os.Getenv(AWSSecretKey) != "") {
```
You can merge this `if` with the next one.

yasker commented on controller/replica_controller.go:
```diff
@@ -389,6 +390,23 @@ func (rc *ReplicaController) CreatePodSpec(obj interface{}) (*v1.Pod, error) {
+			return nil, err
+		}
+		if settings != nil && settings.BackupTargetCredentialSecret != "" {
+			credentialEnv, err := util.ConfigEnvWithCredential(r.Spec.RestoreFrom, settings.BackupTargetCredentialSecret)
```
You can just take the container (`pod.Spec.Containers[0]`) as a parameter and make a function to update its environment.

yasker commented on controller/volume_controller.go:
```diff
@@ -957,6 +957,17 @@ func (vc *VolumeController) createCronJob(v *longhorn.Volume, job *types.Recurri
 		},
 	}
+	if job.Type == types.RecurringJobTypeBackup {
+		credentialEnv, err := util.ConfigEnvWithCredential(backupTarget, credentialSecret)
```
Same here.

yasker commented on util/util.go:
```diff
@@ -26,9 +26,10 @@ const (
+	AWSAccessKey      = "AWS_ACCESS_KEY_ID"
+	AWSSecretKey      = "AWS_SECRET_ACCESS_KEY"
 	BackupStoreTypeS3 = "s3"
+	AWSEndPoint       = "AWS_ENDPOINTS"
```
This looks like a weird sandwich... Can you put `BackupStoreTypeS3` above the `AWSAccessKey`?

yasker commented:
LGTM. Will merge after Longhorn Engine is updated.

yasker commented:
@JacieChao I've merged and updated the engine image for S3. Can you rebase this PR?

JacieChao commented:
@yasker Sure. I will update this PR after final test.

JacieChao commented:
@yasker I have rebased master code, but longhorn-manager running with errors when using latest longhorn-engine image(`cannot get engine version for xxxxxx`) since there's no `version` command for longhorn-engine. 
If there's something wrong with my usage, please let me know, or I can test it again after longhorn-engine update.

yasker commented:
@JacieChao Sorry for the confusion. I forgot to check in one commit of the engine part. It should be fine now.

JacieChao commented:
@yasker OK, I will ping you after final test.

JacieChao commented:
@yasker I have rebased code and test with newest image, longhorn-manager works well with s3 backup. 
I also rebased longhorn-test code and running the test.
```
+ flake8 .
+ py.test -v .
============================= test session starts ==============================
platform linux2 -- Python 2.7.14, pytest-2.9.2, py-1.5.3, pluggy-0.3.1 -- /usr/local/bin/python
cachedir: .cache
rootdir: /integration/tests, inifile:
collecting ... collected 14 items
test_basic.py::test_hosts_and_settings PASSED
test_basic.py::test_volume_basic PASSED
test_basic.py::test_snapshot PASSED
test_basic.py::test_backup PASSED
test_basic.py::test_volume_multinode PASSED
test_csi.py::test_csi_volume_mount SKIPPED
test_csi.py::test_csi_volume_io SKIPPED
test_driver.py::test_volume_mount PASSED
test_driver.py::test_volume_io PASSED
test_engine_upgrade.py::test_engine_offline_upgrade SKIPPED
test_engine_upgrade.py::test_engine_live_upgrade SKIPPED
test_ha.py::test_ha_simple_recovery PASSED
test_ha.py::test_ha_salvage PASSED
test_recurring_job.py::test_recurring_job SKIPPED
=============== 9 passed, 5 skipped in 596.73 seconds ================
```

yasker commented:
Merged, thanks.

Have a good night.
